File: algorithmAndDataStructure/golang/addTwoNumbers/addTwoNumbers.go
package addTwoNumbers

// ListNode is definition for singly-linked list
type ListNode struct {
	Val  int
	Next *ListNode
}

func getValueAndNext(l *ListNode) (next *ListNode, value int) {
	value = 0
	if l != nil {
		value = l.Val
		l = l.Next
	}
	next = l
	return
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	var head *ListNode
	for carry, tmp := 0, &head; l1 != nil || l2 != nil || carry > 0; {
		var x, y int
		l1, x = getValueAndNext(l1)
		l2, y = getValueAndNext(l2)
		sum := x + y + carry
		node := &ListNode{
			Val:  sum % 10,
			Next: nil,
		}
		*tmp = node
		tmp = &node.Next

		carry = sum / 10
	}
	return head
}

File: algorithmAndDataStructure/cpp/addTwoNumbers/addTwoNumbers.cpp
#include <stdio.h>
#include <algorithm>
#include <vector>
#include <iostream>

using namespace std;

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int x = 0 , y = 0, sum = 0, carry=0;
        ListNode *head=NULL, **tmp=&head;
        while(l1 != NULL || l2 != NULL || carry>0) {
            x = getValueAndNext(l1);
            y = getValueAndNext(l2);
            sum = x + y + carry;
            ListNode *node = new ListNode(sum%10);
            *tmp = node;
            tmp = &node->next;

            carry = sum/10;
        }
        return head;
    }
private:
    int getValueAndNext(ListNode* &node) {
        int value=0;
        if(node != NULL) {
            value = node->val;
            node = node->next;
        }
        return value;
    }
};